plugins/pluginManager.js

```diff
@@ -0,0 +1,167 @@
+/**
+ * PluginManager.js
+ *
+ * A bare bones plugin manager.  It is meant to look very much like the linux
+ * process management.  It creates hooks for the startup of plugins, restarts,
+ * and halts.
+ *
+*/
+
+var path =  require('path'),
+    thread = require('child_process');
+
+
+var MAX_RESTARTS = 3;
+
+/*
+ * An instance of a plugin.
+ * @constructor
+ * @param {string} script - the script to execute a node process on.
+ *
+*/
+function Plugin(script) {
+
+  // the location of the program
+  this.script = script;
+
+  // number of attempted restarts
+  this.restarts = 0;
+
+  // the number of times to attempt a restart
+  // in case of unknown termination signal.
+  this.maxRestarts = MAX_RESTARTS;
+
+
+  // perform the initial startup
+  this.startup();
+}
+
+
+/** performs the initial fork, and sets up handlers */
+Plugin.prototype.startup = function () {
+
+  // fork the process and assign it to this.process
+  this.process = thread.fork(__dirname + this.script);
+
+  // set up the exit handler
+  this.process.on('exit', this.exitHandler);
+
+  // the child fork is running
+  this.running = true;
+};
+
+
+/** exit handler for the underlying process */
+Plugin.prototype.exitHandler = function (code, signal) {
+
+  // 0 is the successful exit code
+  // SIGTERM is emitted on process.kill()
+  if (code !== 0 && signal !== 'SIGTERM') {
+
+    // if we have received more than the maxRestarts, something may be amiss
+    // and we should quit trying to restart
+    if (this.restarts < this.maxRestarts) {
+      this.startup();
+      this.restarts++;
+    }
+  }
+};
+
+
+/** nice event emitter for the underlying process */
+Plugin.prototype.emit = function (event, data) {
+  this.process.send({ event : event, data : data });
+};
+
+
+/** wraps process.kill() */
+Plugin.prototype.kill = function (code) {
+  this.process.kill(code);
+};
+
+
+/** wraps process.on() */
+Plugin.prototype.register = function (event, callback) {
```

Comment:
It is being used in a separate branch of mine (plugins/event-monitor), which hasn't been pushed to GitHub yet.  It's part of a feature to monitor the status of plugins from the bhima web application over HTTP.